### PR TITLE
Add support for objective and constraint hessians and `jac=True` option for constraints in the scipy interface

### PR DESCRIPTION
**Pull request by jhelgert**
Implements #99 and #122.

- Adds support for objective hessian and constraint Hessians through an additional `hess` field in the constraint dictionary.  The hessian must have the signature `hess(x,v) -> np.ndarray` and return `v[0] * Hg1_(x) + ... + v[m-1] * Hg_m(x)`. Here, `v` is the np.ndarray with shape `m` containing the lagrangian multipliers for the constraint and `Hg_i` denotes the hessian matrix of the constraint component `g_i : R^N -> R`.

Current limitations:
- No support for NonlinearConstraint objects.
- No support for objective hessian matrix-vector product `hessp`.
- In case one passes the hessian for at least one constraint or the objective, we need the Hessians for all constraints and the objective.

### CODE REVIEW

**Comment by moorepants** [comment]
@jhelgert Is this ready to be reviewed? Whenever you think it is done, let us know.

**Comment by jhelgert** [comment]
@moorepants Yes, I think it's ready. However, there are some unintentional formatting changes because my code formatter seems to ignore the settings :/.

File: cyipopt/scipy_interface.py
Comment (moorepants) [inline cyipopt/scipy_interface.py:74]: Adding the arg `n` is technically backward incompatible since you can supply kwargs in order as args. Can `n` be a kwarg added at the end? Same goes for the order of `con_dims`.

File: cyipopt/scipy_interface.py
Comment (moorepants) [inline cyipopt/scipy_interface.py:178]: Any new public methods should have complete docstrings following the numpydoc format.

File: cyipopt/scipy_interface.py
Comment (moorepants) [inline cyipopt/scipy_interface.py:214]: This is also backwards incompatible. Best to make a new function to get the dimensions.

Comment (moorepants) [reply cyipopt/scipy_interface.py:214]: You could add a kwarg that causes the number of returned item to increase in length, that's ok, but still not as clean as it could be.

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (moorepants) [inline cyipopt/tests/unit/test_scipy_optional.py:79]:
```suggestion
         and constraint jacobians and Hessians"""
```

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (moorepants) [inline cyipopt/tests/unit/test_scipy_optional.py:100]:
```suggestion
         and constraint jacobians and Hessians"""
```

File: cyipopt/scipy_interface.py
Comment (jhelgert) [reply cyipopt/scipy_interface.py:74]: > Can `n` be a kwarg added at the end? Same goes for the order of `con_dims`.

Of course. Just to be sure I don't confuse it with the first `kwargs`: You mean something like

``` python
def __init__(self, fun, args=(), kwargs=None, jac=None, hess=None, hessp=None, constraints=(), eps=1e-8, **kkwargs):
    self.n = kkwargs.get('n', None)
    # ....
    self.con_dims = kkwargs.get('con_dims', [])
    # ....
```
right?

File: cyipopt/scipy_interface.py
Comment (jhelgert) [reply cyipopt/scipy_interface.py:178]: I can't think of a use case in which the user might need this method, so I'd propose to make it private. Moreover, the `i` argument for the i-th constraint doesn't feel pythonic to me, so another reason to make it private. What do you think?

File: cyipopt/scipy_interface.py
Comment (moorepants) [reply cyipopt/scipy_interface.py:74]: Can this work?

```python
def __init__(self, fun, args=(), kwargs=None, jac=None, hess=None, hessp=None, constraints=(), eps=1e-8, n=None:
```

File: cyipopt/scipy_interface.py
Comment (moorepants) [reply cyipopt/scipy_interface.py:178]: Making it private also need deprecation.

Comment (moorepants) [reply cyipopt/scipy_interface.py:178]: Oh, sorry, misunderstood. Yes, make it private.

**Comment by jhelgert** [comment]
@moorepants I think it's ready to be reviewed again. Thanks!

File: examples/hs071_scipy_jax.py
Comment (brocksam) [inline examples/hs071_scipy_jax.py:65]:
```suggestion
print(res)

```

File: examples/hs071_scipy_jax.py
Comment (brocksam) [inline examples/hs071_scipy_jax.py:42]:
```suggestion
obj_hess = jit(jacrev(jacfwd(obj_jit)))  # objective hessian
```

File: examples/hs071_scipy_jax.py
Comment (brocksam) [inline examples/hs071_scipy_jax.py:48]:
```suggestion
con_ineq_hessvp = jit(lambda x, v: con_ineq_hess(x) * v[0])  # hessian vector-product
```

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:106]:
```suggestion
        "jac": True,
```

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:86]: Printing during tests with pytest won't do anything obvious. If this needs to be tested then the stdout can be captured, otherwise this should be removed.
```suggestion
```

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:111]:
```suggestion
```

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:83]: Please keep code style consistent within files.
```suggestion
              "jac": rosen_der, "hess": lambda x, v: rosen_hess(x) * v[0]}
```

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:79]:
```suggestion
    """`minimize_ipopt` works with objective gradient and Hessian 
       and constraint jacobians and Hessians."""
```

File: cyipopt/ipopt_wrapper.py
Comment (brocksam) [inline cyipopt/ipopt_wrapper.py:8]: What's the reasoning for renaming the function and then importing under an alias? Should we just leave the function name as is? Or if the function name needs to chain for clarity reasons but must remain for compatibility, should we support both and deprecate the old name?

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:109]: Is this a public attribute typically supplied as part of the `scipy.optimize.minimize` API? Should this be public or private given that other similar attributes for this class are private?

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:182]: Is double underscore name mangling necessary here? Or is this just intended to be a private method?
```suggestion
    def _evaluate_con_fun_with_jac(self, i, con_fun, x, *args):
```

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:177]: See comment at 182.
```suggestion
                    self._evaluate_con_fun_with_jac(i, fun, x, *args)[1])
```

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:166]: See comment at 182.
```suggestion
                    self._evaluate_con_fun_with_jac(i, fun, x, *args)[0])
```

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:183]: Should this be a `np.isclose` check for `x` rather than a direct check for equality?

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:186]: I personally don't think that it's clear what is being returning whenever this function is called. In the two cases it is called, the return value is index directly. Should this be split in to two separate methods/functions/closures that make the return value obvious?

File: cyipopt/scipy_interface.py
Comment (brocksam) [inline cyipopt/scipy_interface.py:189]:
```suggestion
        return np.nonzero(np.tril(np.ones((self.n, self.n))))  # type: ignore
```

**Review by brocksam — COMMENTED** [review]
Mostly formatting changes, but a couple of questions that might warrant further discussion. Good job! Thanks!

**Comment by moorepants** [comment]
@jhelgert Can you merge in master so the tests at least run on linux?

File: cyipopt/scipy_interface.py
Comment (jhelgert) [reply cyipopt/scipy_interface.py:109]: No, it's only needed to cache the function value and the evaluated Jacobian for each constraint and used inside the `__evaluate_con_fun_with_jac()` method. So I agree that it should rather be private.

File: cyipopt/scipy_interface.py
Comment (jhelgert) [reply cyipopt/scipy_interface.py:183]: It's really intended to be a check for equality since `np.isclose(self.last_x, x)` would return `True` if the new point `x` lies in a small neighborhood  of `self.last_x`. However, we must ensure that it's the same point. It's more or less identical to the `evaluate_fun_with_grad()` method.

File: cyipopt/ipopt_wrapper.py
Comment (jhelgert) [reply cyipopt/ipopt_wrapper.py:8]: Good question. But I'm not quite sure, maybe @moorepants can enlighten us?  Btw: I just noticed it should be `get_constraint_bounds` since I split the function into two smaller ones in one of the previous commits.

File: cyipopt/scipy_interface.py
Comment (jhelgert) [reply cyipopt/scipy_interface.py:186]: I agree it is a bit cumbersome. Does it help to add a few comments? The function returns a tuple `(fun_val, jac_val)` containing the function value `fun_val` and the evaluated jacobian `jac_val` of the `i`-th constraint.

Maybe something like

```  python
def _evaluate_con_fun_with_jac(self, i, con_fun, x, *args):
    if self.last_x is None or not np.all(self.last_x == x):
        self.last_x = x
        self.last_con_fun_values[i], self.last_con_jac_values[i] = con_fun(x, *args)
    return self.last_con_fun_values[i], self.last_con_jac_values[i]
```

makes it easier to understand at the first sight?

**Comment by moorepants** [comment]
I guess appveyor is still running, just not webhooked into here anymore. Appveyor also failed: https://ci.appveyor.com/project/moorepants/cyipopt/builds/40165736

**Comment by jhelgert** [comment]
Thanks to both of you for the reviews so far! I will investigate why the last test `test_minimize_ipopt_constraint_jac_true_if_scipy()` fails on Linux while it passes on macOS. 

PS: Appveyor fails with

``` none
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\projects\cyipopt\cyipopt\__init__.py", line 13, in <module>
    from .ipopt_wrapper import *
  File "C:\projects\cyipopt\cyipopt\ipopt_wrapper.py", line 8, in <module>
    from .scipy_interface import get_constraint_bounds_and_dimensions as cyipopt_get_constraint_bounds
ImportError: cannot import name 'get_constraint_bounds_and_dimensions'
Command exited with code 1
```

This was already fixed with the last commit.

**Comment by jhelgert** [comment]
@moorepants All tests pass now on Linux. However, there's still one Appveyor job that fails:

https://ci.appveyor.com/project/moorepants/cyipopt/builds/40167487

File: cyipopt/scipy_interface.py
Comment (brocksam) [reply cyipopt/scipy_interface.py]: Writing code that is clear in itself is significantly preferable to having to explain the code through comments. As this function is only being called in two places with one accessing the first return value and the other accessing the second, I still believe that the should be split in to two separate methods. One called `_evaluate_last_con_fun_value` and the other `_evaluate_last_con_jac_value` or similar.

**Comment by brocksam** [comment]
> https://ci.appveyor.com/project/moorepants/cyipopt/builds/40167487

Looking at the Appveyor readout it is failing on `assert np.isclose(res.get("fun"), 1.0828541)` because `res.get("fun")` returns 1.003480116709976`. Unclear to me why this would be the case with this specific environment only.

File: cyipopt/scipy_interface.py
Comment (jhelgert) [reply cyipopt/scipy_interface.py]: Since the same argument applies for `evaluate_fun_with_grad()`, shouldn't we then also split this method?

Comment (brocksam) [reply cyipopt/scipy_interface.py]: Always amusing when PR code review flags up inconsistencies elsewhere in your codebase! Truthfully, yes, I think we should also split `evaluate_fun_with_grad()`. Or even better, refactor the relevant parts of it into `objective()` and `gradient()`. However, as `evaluate_fun_with_grad()` has up until this point been a public method, it will need to be deprecated correctly if being removed. In that case, probably best to ignore all of my comments relating to this and instead will open an issue regarding this so it can be fixed in a separate PR. Thanks @jhelgert! :)

**Comment by jhelgert** [comment]
@brocksam Thanks for the discussion! Can you please have another look? I think it should be much clearer now thanks to [`MemoizeJac`](https://github.com/scipy/scipy/blob/47af6102fc5a900c5008bc4ad8415ad2a136c92c/scipy/optimize/optimize.py#L55). And as you already mentioned, we could replace `evaluate_fun_with_grad()` similarly in another PR.

**Review by brocksam — DISMISSED** [review]

**Comment by brocksam** [comment]
> > https://ci.appveyor.com/project/moorepants/cyipopt/builds/40167487
> 
> Looking at the Appveyor readout it is failing on `assert np.isclose(res.get("fun"), 1.0828541)` because `res.get("fun")` returns 1.003480116709976`. Unclear to me why this would be the case with this specific environment only.

@jhelgert this is still a problem.

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:87]: Can you please explain why the expected result of this test is 1.082541 and not 1.0? The problem formulation seems to indicate that the result should in fact be 1.0...

**Review by brocksam — CHANGES_REQUESTED** [review]

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (jhelgert) [reply cyipopt/tests/unit/test_scipy_optional.py:87]: Good catch! Removing the Hessians, the result is 0.999999990217499. After testing a few more test problems, I noticed that there's definitely something wrong with the evaluation of the lagrangian hessian (corresponds to the methods `hessian` and `hessianstructure`) :/. I'll delve into it.

Comment (brocksam) [reply cyipopt/tests/unit/test_scipy_optional.py:87]: Excellent, glad we caught :)

Comment (jhelgert) [reply cyipopt/tests/unit/test_scipy_optional.py:87]: Found it! :)  The formula for the number of non-zero elements in the hessian is wrong, see [here](https://github.com/mechmotum/cyipopt/blob/d1c8d70bcd3c2c186e611f710509d6cc144e123f/cyipopt/cython/ipopt_wrapper.pyx#L366). It should be `n*(n+1)/2` instead of `n*(n-1)/2`. This also (partially) fixes #125. 

Nonetheless, I still obtain an objective value of 1.08 with the 'mu_strategy'='adaptive' for both MWEs. However, both now give the correct solution with objective value 1.0000000552371922 with the default 'mu_strategy' ('monotone').
After having tested some other problems without any errors, I reckon that the above test problem is just a rare corner case in which the different 'mu_strategy' options yield different solutions. So I'd propose to change the above test problem and add some additional test problems (just to play safe). What do you think?

**Comment by jhelgert** [comment]
I think we are getting closer. I removed the moody `test_minimize_ipopt_jac_and_hessians_constraints_monotone_strategy_if_scipy()` test that failed due to different results on macOS/Linux/Windows:

``` python
# macOS: 1.087746, 1.143973, 1.306229, 1.710728, 2.880089
#   Win: 1.087746, 1.143977, 1.306237, 1.710748, 2.880159
# Linux: 1.087745, 1.143968, 1.306218, 1.710698, 2.879984
```
Both `test_minimize_ipopt_jac_and_hessians_constraints_if_scipy()` and `test_minimize_ipopt_hs071()` also use the new constraint Hessians and the tests pass on Linux and Windows (Appveyor).

**Comment by brocksam** [comment]
> I think we are getting closer. I removed the moody `test_minimize_ipopt_jac_and_hessians_constraints_monotone_strategy_if_scipy()` test that failed due to different results on macOS/Linux/Windows:
> 
> ```python
> # macOS: 1.087746, 1.143973, 1.306229, 1.710728, 2.880089
> #   Win: 1.087746, 1.143977, 1.306237, 1.710748, 2.880159
> # Linux: 1.087745, 1.143968, 1.306218, 1.710698, 2.879984
> ```
> 
> Both `test_minimize_ipopt_jac_and_hessians_constraints_if_scipy()` and `test_minimize_ipopt_hs071()` also use the new constraint Hessians and the tests pass on Linux and Windows (Appveyor).

To make sure I've understood this correctly, you're still operating under the assumption that there is a corner case where the different `mu-strategy` options result in different solutions? If this is the case, should we ensure that this is indeed a quirk of Ipopt and not a quirk of CyIpopt by double checking using the C++ API before committing these changes?

**Comment by jhelgert** [comment]
> To make sure I've understood this correctly, you're still operating under the assumption that there is a corner case where the different `mu-strategy` options result in different solutions? 

Yes, exactly. 

> If this is the case, should we ensure that this is indeed a quirk of Ipopt and not a quirk of CyIpopt by double checking using the C++ API before committing these changes?

Before writing it in C++, I have done it quickly with [CasADi](https://github.com/casadi/casadi) which computes all gradients, jacobians and Hessians by algorithmic differentiation and interfaces different solvers including Ipopt. Here, I can reproduce the problem:

``` python3
import casadi as cas

opti = cas.Opti()
x = opti.variable(5)
rosen = sum(100*(x[i+1]-x[i]**2)**2 + (1-x[i])**2 for i in range(4))
opti.minimize(rosen)
opti.subject_to(rosen - 1.0 >= 0)

opti.solver("ipopt", {}, {"mu_strategy": "adaptive"})
opti.set_initial(x, [1.3, 0.7, 0.8, 1.9, 1.2])
sol = opti.solve()

print(sol.value(x))
print(sol.value(rosen))
```

which gives me

``` none
[1.09022019 1.14568612 1.30965198 1.71962945 2.90683522]
1.0828541437101658
```

Using the default 'mu_strategy' the objective value is again 1.0000000578559098 as expected. Therefore, it should indeed be a quirk of Ipopt. Should I still double-check it again with Ipopt's C++ API?

**Comment by brocksam** [comment]
Excellent idea to use CasADi! As you've confirmed it's an Ipopt quirk independent of CyIpopt, that's definitely sufficient.

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:114]: Can you please refactor this docstring so that it's in standard Numpy format? Single line initial description, followed by anything else as the main body etc.

**Review by brocksam — COMMENTED** [review]
Other than the docstring on the new test, I'm very happy with this. Thanks so much for all of your work! I'll defer to @moorepants to give final approval and make the merge.

File: cyipopt/tests/unit/test_scipy_optional.py
Comment (jhelgert) [reply cyipopt/tests/unit/test_scipy_optional.py]: Done.

**Comment by moorepants** [comment]
I'm on vacation right now and haven't had time to do a final check. If Sam has time for a final review and merge, he can do that. Otherwise, I'll try to review in the next week if I can find a bit of time.

**Pull request merged**
